### ISSUE TITLE
Coredns is in running state but not functioning properly as expected.

### ISSUE THREAD

**Issue by panchm** [miekg/coredns#5862]
We are observing multiple issues with the coredns on the below mentioned environment

**Environment**:
- AWS Region: us-east-1
- Instance Type(s): m5.4xlarge and m5.2xlarge
- EKS Platform version: eks.5
- Kubernetes version: v1.23.7
- AMI Version: (ami-036b6ef0ea363fec1) ubuntu-eks/k8s_1.23/images/hvm-ssd/ubuntu-focal-20.04-amd64-server-20220901 
- Kernel : Linux ip-10-0-1-176 5.15.0-1019-aws #23~20.04.1-Ubuntu SMP Thu Aug 18 03:20:14 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux
- Release information (run `cat /etc/eks/release` on a node): Not found
- [Amazon VPC CNI]
 Enable pod networking within your cluster.
Category: networking
Status: Active
Version: v1.10.4-eksbuild.1

- [CoreDNS] Enable service discovery within your cluster.
Category: networking
Status: Active
Version: v1.8.7-eksbuild.2
IAM Role
Inherited from nodeEnable service discovery within your cluster.

- [kube-proxy] Enable service networking within your cluster.
Category: networking
Status: Active
Version: v1.23.7-eksbuild.1
IAM Role
Inherited from node

### 3. We observed one strange behavior i.e. sometimes application pods are running properly if the coredns are not running part of that same node.

We have tried to follow the below steps as well, but does not help much and it helps sometime with restarted the coredns pods.
https://aws.amazon.com/premiumsupport/knowledge-center/eks-dns-failure/

Note: We took precautions with the coredns pods running on different nodes all the time.

Is this a known issue? We have not altered any configuration with respect to the Kube-system pods.
Are we missing some other configuration required as part of the EKS nodegroup or any other configuration?

Sometimes we don't see any errors but the application pods failed to communicate

```
$ kubectl logs --follow -n kube-system --selector 'k8s-app=kube-dns'
.:53
[INFO] plugin/reload: Running configuration MD5 = 47d57903c0f0ba4ee0626a17181e5d94
CoreDNS-1.8.7
linux/amd64, go1.17.7, d433a3f2
.:53
[INFO] plugin/reload: Running configuration MD5 = 47d57903c0f0ba4ee0626a17181e5d94
CoreDNS-1.8.7
linux/amd64, go1.17.7, d433a3f2
```

Sometimes we observe these issues which says some plugin errors
 ```
kubectl logs --follow -n kube-system --selector 'k8s-app=kube-dns'
.:53
[INFO] plugin/reload: Running configuration MD5 = 47d57903c0f0ba4ee0626a17181e5d94
CoreDNS-1.8.7
linux/amd64, go1.17.7, d433a3f2
[ERROR] plugin/errors: 2 sts.us-east-1.amazonaws.com. A: read udp 10.0.173.102:39671->10.0.0.2:53: i/o timeout
[ERROR] plugin/errors: 2 sts.us-east-1.amazonaws.com. A: read udp 10.0.173.102:33958->10.0.0.2:53: i/o timeout
[ERROR] plugin/errors: 2 5155737311967868687.6477139177168497870. HINFO: read udp 10.0.173.102:42392->10.0.0.2:53: i/o timeout
[ERROR] plugin/errors: 2 5155737311967868687.6477139177168497870. HINFO: read udp 10.0.173.102:39358->10.0.0.2:53: i/o timeout
[ERROR] plugin/errors: 2 5155737311967868687.6477139177168497870. HINFO: read udp 10.0.173.102:60994->10.0.0.2:53: i/o timeout
[ERROR] plugin/errors: 2 5155737311967868687.6477139177168497870. HINFO: read udp 10.0.173.102:59687->10.0.0.2:53: i/o timeout
[ERROR] plugin/errors: 2 kafka.kafka-nextgen-dev1.svc.cluster.local.ec2.internal. AAAA: read udp 10.0.173.102:34256->10.0.0.2:53: i/o timeout
[ERROR] plugin/errors: 2 5155737311967868687.6477139177168497870. HINFO: read udp 10.0.173.102:33075->10.0.0.2:53: i/o timeout
[ERROR] plugin/errors: 2 kafka.kafka-nextgen-dev1.svc.cluster.local.ec2.internal. AAAA: read udp 10.0.173.102:37118->10.0.0.2:53: i/o timeout
[ERROR] plugin/errors: 2 5155737311967868687.6477139177168497870. HINFO: read udp 10.0.173.102:35190->10.0.0.2:53: i/o timeout
[ERROR] plugin/errors: 2 5155737311967868687.6477139177168497870. HINFO: read udp 10.0.173.102:53705->10.0.0.2:53: i/o timeout
[ERROR] plugin/errors: 2 5155737311967868687.6477139177168497870. HINFO: read udp 10.0.173.102:33519->10.0.0.2:53: i/o timeout
[ERROR] plugin/errors: 2 5155737311967868687.6477139177168497870. HINFO: read udp 10.0.173.102:52150->10.0.0.2:53: i/o timeout
[ERROR] plugin/errors: 2 sts.us-east-1.amazonaws.com.ec2.internal. A: read udp 10.0.173.102:45225->10.0.0.2:53: i/o timeout
[ERROR] plugin/errors: 2 sts.us-east-1.amazonaws.com.ec2.internal. AAAA: read udp 10.0.173.102:44717->10.0.0.2:53: i/o timeout
[ERROR] plugin/errors: 2 sts.us-east-1.amazonaws.com.ec2.internal. AAAA: read udp 10.0.173.102:50949->10.0.0.2:53: i/o timeout
[ERROR] plugin/errors: 2 sts.us-east-1.amazonaws.com.ec2.internal. A: read udp 10.0.173.102:52404->10.0.0.2:53: i/o timeout
[ERROR] plugin/errors: 2 sts.us-east-1.amazonaws.com. A: read udp 10.0.173.102:47961->10.0.0.2:53: i/o timeout
[ERROR] plugin/errors: 2 sts.us-east-1.amazonaws.com. AAAA: read udp 10.0.173.102:42074->10.0.0.2:53: i/o timeout
```

**Comment by chrisohaver** [comment]
This looks more like a network layer problem than a DNS service problem. Eg packets not reaching CoreDNS, and network failing to transmit packets from coredns to upstream dns server/s.

**Comment by panchm** [comment]
We have investigated the security groups as well on the EC2 instance where already TCP & UDP port 53 is whitelisted

**Comment by chrisohaver** [comment]
As for requests packets that are not delivered to coredns, coredns of course cannot do anything about that or even report that they never arrived (hence no errors in log).  The request is lost before it is received by coredns.

For the timeout errors in the logs, coredns sends a request to an upstream dns server, and never gets a reply. Once the request is sent upstream, it's out of the hands of coredns until a response is received.  In this case the response is never received.

Neither of those symptoms would be due to a _coredns_ failure.  These types of errors are due to network transport issues.
I recommend you troubleshoot the CNI network plugin you are using in your Kubernetes cluster.  Beyond looking at CNI logs, I cannot advise precisely how to troubleshoot the CNI since it's not my area of expertise.

**Comment by kumarpmd** [comment]
@panchm I notice you are on v1.23.7 but not sure about the container runtime.  That said, the error seems similar to what we faced for a very similar configuration.
When we changed container-runtime to containerd`, CoreDNS pods reported timeouts, and other pods failed to initialize   

In our case, system hardening sets net.ipv4.ip_forward to 0.  This was resulting in lost packets and not reach CoreDNS.  
Setting `net.ipv4.ip_forward = 1` enabled CoreDNS to function properly (no timeout errors), and subsequently other pods to initialize and create endpoints..
Run `sudo sysctl net.ipv4.ip_forward=1` and test a deployment.
To fix it permanently - update `net.ipv4.ip_forward=1` in /etc/sysctl.conf
